In [246]:
import gpxpy
import numpy as np
import pandas as pd
import os

In [253]:
# Prepare data frames
df = pd.DataFrame(columns=['file_name','to','fr','diff'])
df1 = pd.DataFrame(columns=['file_name','to','fr','diff'])

In [254]:
gpx_file_list = os.listdir('activities')

In [255]:
for gpx_file_name in gpx_file_list[0:15]:
    # Read GPX file
    gpx_file = open('activities/' + gpx_file_name, 'r')
    gpx = gpxpy.parse(gpx_file)
    start_time, end_time = gpx.get_time_bounds()
    
    # Append GPX file to current data frame
    df1['file_name'] = [gpx_file_name]
    df1['to'] = [start_time]
    df1['fr'] = [end_time]
    df1['diff'] = [0]
    
    df = df.append(df1, ignore_index=True)

In [256]:
df

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>file_name</th>
      <th>to</th>
      <th>fr</th>
      <th>diff</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20180122-173654-Ride.gpx</td>
      <td>2018-01-22 17:36:54</td>
      <td>2018-01-22 18:28:27</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>20171207-014844-Ride.gpx</td>
      <td>2017-12-07 01:48:44</td>
      <td>2017-12-07 02:55:30</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20170808-150751-Ride.gpx</td>
      <td>2017-08-08 15:07:51</td>
      <td>2017-08-08 16:01:51</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>20180116-154212-Ride.gpx</td>
      <td>2018-01-16 15:42:12</td>
      <td>2018-01-16 16:32:55</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>20171012-235613-Ride.gpx</td>
      <td>2017-10-12 23:56:13</td>
      <td>2017-10-13 00:51:59</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>20171215-225105-Ride.gpx</td>
      <td>2017-12-15 22:51:05</td>
      <td>2017-12-15 23:49:27</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>20171208-235221-Ride.gpx</td>
      <td>2017-12-08 23:52:21</td>
      <td>2017-12-09 00:52:18</td>
      <td>0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>20180205-154548-Ride.gpx</td>
      <td>2018-02-05 15:45:48</td>
      <td>2018-02-05 16:43:29</td>
      <td>0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>20170828-235331-Ride.gpx</td>
      <td>2017-08-28 23:53:31</td>
      <td>2017-08-29 00:51:51</td>
      <td>0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>20180115-160656-Ride.gpx</td>
      <td>2018-01-15 16:06:56</td>
      <td>2018-01-15 16:59:18</td>
      <td>0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>20170724-141036-Ride.gpx</td>
      <td>2017-07-24 14:10:36</td>
      <td>2017-07-24 14:26:51</td>
      <td>0</td>
    </tr>
    <tr>
      <th>11</th>
      <td>20180208-020057-Ride.gpx</td>
      <td>2018-02-08 02:00:57</td>
      <td>2018-02-08 02:55:11</td>
      <td>0</td>
    </tr>
    <tr>
      <th>12</th>
      <td>20180129-152408-Ride.gpx</td>
      <td>2018-01-29 15:24:08</td>
      <td>2018-01-29 16:29:47</td>
      <td>0</td>
    </tr>
    <tr>
      <th>13</th>
      <td>20170831-133621-Ride.gpx</td>
      <td>2017-08-31 13:36:21</td>
      <td>2017-08-31 14:20:45</td>
      <td>0</td>
    </tr>
    <tr>
      <th>14</th>
      <td>20180206-153558-Ride.gpx</td>
      <td>2018-02-06 15:35:58</td>
      <td>2018-02-06 15:44:12</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>

In [257]:
# Convert timezones
df['to'] = df.to.apply(lambda x: x.tz_localize('UTC').tz_convert('America/Los_Angeles'))
df['fr'] = df.fr.apply(lambda x: x.tz_localize('UTC').tz_convert('America/Los_Angeles'))

In [258]:
# Calculate time difference
df['diff'] = (df['fr'] - df['to'])
df

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>file_name</th>
      <th>to</th>
      <th>fr</th>
      <th>diff</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20180122-173654-Ride.gpx</td>
      <td>2018-01-22 09:36:54-08:00</td>
      <td>2018-01-22 10:28:27-08:00</td>
      <td>00:51:33</td>
    </tr>
    <tr>
      <th>1</th>
      <td>20171207-014844-Ride.gpx</td>
      <td>2017-12-06 17:48:44-08:00</td>
      <td>2017-12-06 18:55:30-08:00</td>
      <td>01:06:46</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20170808-150751-Ride.gpx</td>
      <td>2017-08-08 08:07:51-07:00</td>
      <td>2017-08-08 09:01:51-07:00</td>
      <td>00:54:00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>20180116-154212-Ride.gpx</td>
      <td>2018-01-16 07:42:12-08:00</td>
      <td>2018-01-16 08:32:55-08:00</td>
      <td>00:50:43</td>
    </tr>
    <tr>
      <th>4</th>
      <td>20171012-235613-Ride.gpx</td>
      <td>2017-10-12 16:56:13-07:00</td>
      <td>2017-10-12 17:51:59-07:00</td>
      <td>00:55:46</td>
    </tr>
    <tr>
      <th>5</th>
      <td>20171215-225105-Ride.gpx</td>
      <td>2017-12-15 14:51:05-08:00</td>
      <td>2017-12-15 15:49:27-08:00</td>
      <td>00:58:22</td>
    </tr>
    <tr>
      <th>6</th>
      <td>20171208-235221-Ride.gpx</td>
      <td>2017-12-08 15:52:21-08:00</td>
      <td>2017-12-08 16:52:18-08:00</td>
      <td>00:59:57</td>
    </tr>
    <tr>
      <th>7</th>
      <td>20180205-154548-Ride.gpx</td>
      <td>2018-02-05 07:45:48-08:00</td>
      <td>2018-02-05 08:43:29-08:00</td>
      <td>00:57:41</td>
    </tr>
    <tr>
      <th>8</th>
      <td>20170828-235331-Ride.gpx</td>
      <td>2017-08-28 16:53:31-07:00</td>
      <td>2017-08-28 17:51:51-07:00</td>
      <td>00:58:20</td>
    </tr>
    <tr>
      <th>9</th>
      <td>20180115-160656-Ride.gpx</td>
      <td>2018-01-15 08:06:56-08:00</td>
      <td>2018-01-15 08:59:18-08:00</td>
      <td>00:52:22</td>
    </tr>
    <tr>
      <th>10</th>
      <td>20170724-141036-Ride.gpx</td>
      <td>2017-07-24 07:10:36-07:00</td>
      <td>2017-07-24 07:26:51-07:00</td>
      <td>00:16:15</td>
    </tr>
    <tr>
      <th>11</th>
      <td>20180208-020057-Ride.gpx</td>
      <td>2018-02-07 18:00:57-08:00</td>
      <td>2018-02-07 18:55:11-08:00</td>
      <td>00:54:14</td>
    </tr>
    <tr>
      <th>12</th>
      <td>20180129-152408-Ride.gpx</td>
      <td>2018-01-29 07:24:08-08:00</td>
      <td>2018-01-29 08:29:47-08:00</td>
      <td>01:05:39</td>
    </tr>
    <tr>
      <th>13</th>
      <td>20170831-133621-Ride.gpx</td>
      <td>2017-08-31 06:36:21-07:00</td>
      <td>2017-08-31 07:20:45-07:00</td>
      <td>00:44:24</td>
    </tr>
    <tr>
      <th>14</th>
      <td>20180206-153558-Ride.gpx</td>
      <td>2018-02-06 07:35:58-08:00</td>
      <td>2018-02-06 07:44:12-08:00</td>
      <td>00:08:14</td>
    </tr>
  </tbody>
</table>
</div>

In [326]:
df = pd.DataFrame(columns=['start','id','elapsed_time'])
df1 = pd.DataFrame(columns=['start','id','elapsed_time'])

In [434]:
#############

from stravalib.client import Client
import json

with open('config.json') as data_file:    
    passwords = json.load(data_file)

client = Client()
access_token = passwords['passwords'][0]['access_token']
client_id = passwords['passwords'][1]['client_id']
client_secret = passwords['passwords'][2]['client_secret']

# Now store that access token somewhere (a database?)
client.access_token = access_token

for activity in client.get_activities():
    df1['start'] = [activity.start_date_local]
    df1['id'] = [activity.id]
    df1['elapsed_time'] = [activity.elapsed_time]
    df = df.append(df1, ignore_index=True)

In [435]:
df

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>id</th>
      <th>elapsed_time</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2018-02-09 07:24:40</td>
      <td>1399201204</td>
      <td>00:19:14</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2018-02-08 17:42:23</td>
      <td>1398534051</td>
      <td>00:59:26</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2018-02-08 06:57:46</td>
      <td>1397660463</td>
      <td>00:51:14</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2018-02-07 18:00:57</td>
      <td>1397004608</td>
      <td>00:58:05</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2018-02-07 08:03:05</td>
      <td>1396154091</td>
      <td>00:17:10</td>
    </tr>
    <tr>
      <th>5</th>
      <td>2018-02-07 07:36:25</td>
      <td>1396104310</td>
      <td>00:08:36</td>
    </tr>
    <tr>
      <th>6</th>
      <td>2018-02-06 15:38:06</td>
      <td>1395231951</td>
      <td>00:10:02</td>
    </tr>
    <tr>
      <th>7</th>
      <td>2018-02-06 15:00:10</td>
      <td>1395194083</td>
      <td>00:13:53</td>
    </tr>
    <tr>
      <th>8</th>
      <td>2018-02-06 08:03:47</td>
      <td>1394477096</td>
      <td>00:12:25</td>
    </tr>
    <tr>
      <th>9</th>
      <td>2018-02-06 07:35:58</td>
      <td>1394431168</td>
      <td>00:08:31</td>
    </tr>
    <tr>
      <th>10</th>
      <td>2018-02-05 17:50:36</td>
      <td>1393732096</td>
      <td>00:57:35</td>
    </tr>
    <tr>
      <th>11</th>
      <td>2018-02-05 07:45:48</td>
      <td>1392985979</td>
      <td>01:00:30</td>
    </tr>
    <tr>
      <th>12</th>
      <td>2018-02-03 17:27:35</td>
      <td>1390461686</td>
      <td>00:26:29</td>
    </tr>
    <tr>
      <th>13</th>
      <td>2018-02-03 17:18:04</td>
      <td>1390362682</td>
      <td>00:07:06</td>
    </tr>
    <tr>
      <th>14</th>
      <td>2018-02-03 17:11:13</td>
      <td>1390331203</td>
      <td>00:06:58</td>
    </tr>
    <tr>
      <th>15</th>
      <td>2018-02-02 17:08:39</td>
      <td>1390331206</td>
      <td>00:16:42</td>
    </tr>
    <tr>
      <th>16</th>
      <td>2018-02-02 16:34:40</td>
      <td>1388559935</td>
      <td>00:20:56</td>
    </tr>
    <tr>
      <th>17</th>
      <td>2018-02-02 08:10:58</td>
      <td>1387982305</td>
      <td>00:18:30</td>
    </tr>
    <tr>
      <th>18</th>
      <td>2018-02-02 07:42:02</td>
      <td>1387939679</td>
      <td>00:09:26</td>
    </tr>
    <tr>
      <th>19</th>
      <td>2018-02-01 17:38:01</td>
      <td>1387244323</td>
      <td>00:11:21</td>
    </tr>
    <tr>
      <th>20</th>
      <td>2018-02-01 16:55:52</td>
      <td>1387229777</td>
      <td>00:15:58</td>
    </tr>
    <tr>
      <th>21</th>
      <td>2018-02-01 07:19:10</td>
      <td>1386505303</td>
      <td>01:59:23</td>
    </tr>
    <tr>
      <th>22</th>
      <td>2018-01-30 06:44:48</td>
      <td>1382997692</td>
      <td>00:48:30</td>
    </tr>
    <tr>
      <th>23</th>
      <td>2018-01-29 19:24:43</td>
      <td>1382306203</td>
      <td>00:50:20</td>
    </tr>
    <tr>
      <th>24</th>
      <td>2018-01-29 07:24:08</td>
      <td>1381583880</td>
      <td>01:07:11</td>
    </tr>
    <tr>
      <th>25</th>
      <td>2018-01-22 09:36:54</td>
      <td>1380265019</td>
      <td>00:56:22</td>
    </tr>
    <tr>
      <th>26</th>
      <td>2018-01-18 18:04:19</td>
      <td>1364054295</td>
      <td>00:22:06</td>
    </tr>
    <tr>
      <th>27</th>
      <td>2018-01-18 17:20:01</td>
      <td>1364025995</td>
      <td>00:29:32</td>
    </tr>
    <tr>
      <th>28</th>
      <td>2018-01-18 08:02:43</td>
      <td>1363524278</td>
      <td>01:06:08</td>
    </tr>
    <tr>
      <th>29</th>
      <td>201

In [411]:
thresh_upper = np.timedelta64(100,"m")
thresh_lower = np.timedelta64(30,"m")

mask = df['elapsed_time'][20:].apply(lambda x: True if (x > thresh_lower and x < thresh_upper) else False)

filtered_array = df['elapsed_time'][20:].where(mask)

filtered_array.mean()

Timedelta('0 days 00:59:33.470238')

In [412]:
filtered_array.size - np.count_nonzero(np.isnat(filtered_array))

168

### Average commute time using CalTrain is 1:00 using 168 commutes for data